File: hydra-java/src/main/java/hydra/impl/java/dsl/LiteralTypes.java
package hydra.impl.java.dsl;

import hydra.core.FloatType;
import hydra.core.IntegerType;
import hydra.core.LiteralType;


public interface LiteralTypes {
  static LiteralType bigfloat() {
    return float_(new FloatType.Bigfloat());
  }

  static LiteralType bigint() {
    return integer(new IntegerType.Bigint());
  }

  static LiteralType binary() {
    return new LiteralType.Binary();
  }

  static LiteralType boolean_() {
    return new LiteralType.Boolean_();
  }

  static LiteralType float32() {
    return float_(new FloatType.Float32());
  }

  static LiteralType float64() {
    return float_(new FloatType.Float64());
  }

  static LiteralType float_(final FloatType ftype) {
    return new LiteralType.Float_(ftype);
  }

  static LiteralType int16() {
    return integer(new IntegerType.Int16());
  }

  static LiteralType int32() {
    return integer(new IntegerType.Int32());
  }

  static LiteralType int64() {
    return integer(new IntegerType.Int64());
  }

  static LiteralType int8() {
    return integer(new IntegerType.Int8());
  }

  static LiteralType integer(final IntegerType itype) {
    return new LiteralType.Integer_(itype);
  }

  static LiteralType uint16() {
    return integer(new IntegerType.Uint16());
  }

  static LiteralType uint32() {
    return integer(new IntegerType.Uint32());
  }

  static LiteralType uint64() {
    return integer(new IntegerType.Uint64());
  }

  static LiteralType uint8() {
    return integer(new IntegerType.Uint8());
  }
}

File: hydra-python/src/main/python/hydra/dsl/literal_types.py
"""A DSL of short names for literal types."""

from hydra.core import (
    FloatType,
    LiteralType,
    LiteralTypeFloat,
    LiteralTypeInteger,
    IntegerType,
    LiteralTypeBinary,
    LiteralTypeBoolean,
    LiteralTypeString,
)


def binary() -> LiteralType:
    """Construct a binary literal type."""
    return LiteralTypeBinary(None)


def boolean() -> LiteralType:
    """Construct a boolean literal type."""
    return LiteralTypeBoolean(None)


def string() -> LiteralType:
    """Construct a string literal type."""
    return LiteralTypeString(None)


def float_(ftype: FloatType) -> LiteralType:
    """Construct a float literal type."""
    return LiteralTypeFloat(ftype)


def float32() -> LiteralType:
    """Construct a float32 literal type."""
    return float_(FloatType.FLOAT32)


def float64() -> LiteralType:
    """Construct a float64 literal type."""
    return float_(FloatType.FLOAT64)


def bigfloat() -> LiteralType:
    """Construct a bigfloat literal type."""
    return float_(FloatType.BIGFLOAT)


def integer(itype: IntegerType) -> LiteralType:
    """Construct an integer literal type."""
    return LiteralTypeInteger(itype)


def int8() -> LiteralType:
    """Construct an int8 literal type."""
    return integer(IntegerType.INT8)


def int16() -> LiteralType:
    """Construct an int16 literal type."""
    return integer(IntegerType.INT16)


def int32() -> LiteralType:
    """Construct an int32 literal type."""
    return integer(IntegerType.INT32)


def int64() -> LiteralType:
    """Construct an int64 literal type."""
    return integer(IntegerType.INT64)


def bigint() -> LiteralType:
    """Construct a bigint literal type."""
    return integer(IntegerType.BIGINT)


def uint8() -> LiteralType:
    """Construct a uint8 literal type."""
    return integer(IntegerType.UINT8)


def uint16() -> LiteralType:
    """Construct a uint16 literal type."""
    return integer(IntegerType.UINT16)


def uint32() -> LiteralType:
    """Construct a uint32 literal type."""
    return integer(IntegerType.UINT32)


def uint64() -> LiteralType:
    """Construct a uint64 literal type."""
    return integer(IntegerType.UINT64)


def show_literal_type(ltype: LiteralType) -> str:
    """Show a literal type as a string."""
    match ltype:
        case LiteralTypeBinary():
            return "binary"
        case LiteralTypeBoolean():
            return "boolean"
        case LiteralTypeFloat(value):
            match value:
                case FloatType.BIGFLOAT:
                    return "float:bigfloat"
                case FloatType.FLOAT32:
                    return "float:float32"
                case FloatType.FLOAT64:
                    return "float:float64"
        case LiteralTypeInteger(value):
            match value:
                case IntegerType.BIGINT:
                    return "integer:bigint"
                case IntegerType.INT8:
                    return "integer:int8"
                case IntegerType.INT16:
                    return "integer:int16"
                case IntegerType.INT32:
                    return "integer:int32"
                case IntegerType.INT64:
                    return "integer:int64"
                case IntegerType.UINT8:
                    return "integer:uint8"
                case IntegerType.UINT16:
                    return "integer:uint16"
                case IntegerType.UINT32:
                    return "integer:uint32"
                case IntegerType.UINT64:
                    return "integer:uint64"
        case LiteralTypeString():
            return "string"
